File: src/main/java/liqp/filters/LFilter.java
package liqp.filters;

import java.util.List;
import java.util.concurrent.atomic.AtomicReference;
import liqp.nodes.LNode;
import liqp.nodes.RenderContext;

/**
 * Interface for performing filter lifecycle operations.
 */
public interface LFilter {

  String getName();

  void doFilter(FilterParams params,
                FilterChainPointer chain,
                RenderContext context,
                AtomicReference<Object> result);

  default void onStartChain(FilterParams params,
                            FilterChainPointer chain,
                            RenderContext context,
                            AtomicReference<Object> result) {

  }

  default void onFilterAction(FilterParams params,
                              FilterChainPointer chain,
                              RenderContext context,
                              AtomicReference<Object> result) {

  }

  default void onEndChain(FilterParams params,
                          FilterChainPointer chain,
                          RenderContext context,
                          AtomicReference<Object> result) {

  }
}

File: src/main/kotlin/liqp/filters/LFilter.kt
package liqp.filters

import liqp.nodes.RenderContext
import java.util.concurrent.atomic.AtomicReference

/**
 * Interface for performing filter operations.
 */
interface LFilter {
  fun doFilter(params: FilterParams,
               chain: FilterChainPointer,
               context: RenderContext,
               result: AtomicReference<Any?>)

  fun doStartFilterChain(params: FilterParams,
                         chain: FilterChainPointer,
                         context: RenderContext,
                         result: AtomicReference<Any?>)

  fun doFilterAction(params: FilterParams,
                     chain: FilterChainPointer,
                     context: RenderContext,
                     result: AtomicReference<Any?>)

  fun doEndFilterChain(params: FilterParams,
                       chain: FilterChainPointer,
                       context: RenderContext,
                       result: AtomicReference<Any?>)
}
